feat: add support for XINFO STREAM stream_name

rounaknandanwar opened the pull request:
Fixes #1517 

<!--
**Commits Must Be Signed and Your PR title must conform to the conventional commit spec**
  * See: https://github.com/dragonflydb/dragonfly/blob/main/CONTRIBUTING.md
  * Please follow the section on `pre-commit hooks`, a linter will validate before you push

  Example PR Title: <type>(<scope>)!: <description>

  * `type` = bug, chore, feat, fix, docs, build, style, refactor, perf, test
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs"
  * `description` = short description of the change

Examples:

  * chore(examples): Clarify `docker` usage #120
  * docs(readme): Fix Example Links #121
  * feat(ingest)!: Add new ingest #122
  * fix(ingest): Refactor for loop to list comprehension #123
-->

kostasrim commented:
Hi @rounaknandanwar may I ask what the status of this is?

rounaknandanwar commented:
> Hi @rounaknandanwar may I ask what the status of this is?

Hi @kostasrim , Got caught up at work.
I am planning to complete it this weekend.

rounaknandanwar commented on src/redis/stream.h:
```diff
@@ -170,7 +170,7 @@ robj *createObjectFromStreamID(streamID *id);
 int streamDeleteItem(stream *s, streamID *id);
 void streamGetEdgeID(stream *s, int first, int skip_tombstones, streamID *edge_id);
-long long streamEstimateDistanceFromFirstEverEntry(stream *s, streamID *id);
+// long long streamEstimateDistanceFromFirstEverEntry(stream *s, streamID *id);
```
@romange Will it be okay to use this method implementation from the original redis source code ? I can see it as commented out right now.
I have replicated it in `stream_family.cc` for now.

romange replied:
it's totally fine reusing the existing functions under src/redis instead of copying code to stream_family if it's possible.

romange replied:
looked at it again. Please uncomment it in t_stream.cc. 

@kostasrim  I am taking ownership of this review. @rounaknandanwar  Please tell me if you want to proceed.

kostasrim replied:
@romange shall I still go over it or?

rounaknandanwar commented:
@kostasrim Can you start reviewing the code? I am fixing a couple of TODOs and adding Unit Tests.
Maybe you can go over the overall structure and flow.

rounaknandanwar commented on src/server/stream_family.cc:
```diff
@@ -843,6 +879,296 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+  ginfo->consumer_info_vec = consumer_info_vec;
+}
+
+OpResult<StreamInfo> OpStreams(const DbContext& db_cntx, string_view key, EngineShard* shard,
```
This function is called from the main xinfo command function. 
This function builds the StreamInfo object which contains all the info that needs to be output.
I have kept the building logic and output print logic separate.

kostasrim reviewed (commented):
Hello @rounaknandanwar, somehow I missed that you opened the PR but I was able to go over it today. 

Thank you again for the contribution! 

I pulled your branch and played around a little bit. There are a few issues:

I created the following stream/group/consumer in df:

```
XADD mystream 1-1 message "Hello,"
XADD mystream 1-2 message " World!"
XGROUP CREATE mystream mygroup $ MKSTREAM
XGROUP CREATECONSUMER mystream mygroup kostas-consumer
```

Then I tested with this commands:
`xinfo help` 
All good, missing the help field in the end (but that's not mandatory although a welcome addition) 

`xinfo consumers mystream mygroup` 
Does not work. Returns:  
`ERR Unknown subcommand or wrong number of arguments for 'CONSUMERS'. Try XINFO HELP.`

2. I also run the commands `xinfo groups mystream` and it has some missing fields like `entries-read`. (compare that with redis output to see the rest)

3. `xinfo stream mystream FULL`
Missing active time, not sure if we track this
>> `(error) ERR unknown subcommand or wrong number of arguments for 'STREAM'. Try XINFO HELP.`

4. `xinfo stream mystream FULL COUNT 0` 


5. Could you also plz add tests?

p.s. I will do a detailed review on the code once these are fixed since I don't want to comment on staff that you might fix/change

rounaknandanwar commented:
Hi @kostasrim thanks for taking a look.

0. I will add the help for `xinfo stream`
1. Checked `xinfo consumers mystream mygroup` on main branch and returns the same error message as you mentioned. I haven't touched the consumers part of xinfo. Do you want me to handle that in this change or some other PR? 
2. For `xinfo groups mystream` command, should it be done in this change or a separate one ? I haven't touched the other parts of xinfo apart from xinfo stream yet hence this is missing.
3. I couldn't find an active-time field on the xinfo stream command. I am using the redis version `7.0.4`. 
4. 0 means all entries to be displayed. I will fix this.
5. I will be adding tests in a couple of days.

A general question on the xinfo command - should I take up the changes on just the stream command or the complete xinfo command(related to points 1 & 2)

Again, apologies for the delay on this.

kostasrim commented:
Hi @rounaknandanwar 

1. No rush, take your time and just let me know when you address everything && I will take a look
2. If you can cover everything in one PR I would be really happy since I can review everything in one go. However, if you think it's too much just let me know and I will take care of the rest.

Ping me/rerequest for review when you cover the list :) 

Again, thank you very much for your contribution :pray:

rounaknandanwar commented:
Hello @kostasrim,

1. I have covered all the sub-commands of xinfo in this PR.
2. I have added tests as well. This should give you a fair idea about the cases being handled.
3. Noticed that `active-time` is not maintained, so haven't added that.
4. There is an issue with entries_read and lag, need help here.

rounaknandanwar commented on src/server/stream_family.cc:
```diff
@@ -836,6 +1005,11 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
       ginfo.consumer_size = raxSize(cg->consumers);
       ginfo.pending_size = raxSize(cg->pel);
       ginfo.last_id = cg->last_id;
+      // ginfo.entries_read = cg->entries_read;  // TODO : this returns incorrect value, check.
```
`cg->entries_read` returns entries_read as 0 after the messages are read using xreadgroup.
I have traced the implementation of xread and xreadgroup but can't see entries_read being updated anywhere.
@kostasrim Can you help confirm this ?

rounaknandanwar replied:
Debugged the xread command and noticed we do not update the `entries_read` field.
`OpRead` function makes use of `OpRangeFromConsumerPEL` & `OpRange`. Both should not modify the `entries_read` field for a group.
It would make sense to update `entries_read` after these calls.
What do you think ?
Also, it would make sense to fix this in a separate PR.

romange replied:
Totally. Please try making this PR as small and self contained as possible even if we miss some semantics. This change is already big as it is. If you find more such issues - you can just add a `TODO: ` comment or open an issue in the repo

rounaknandanwar replied:
Created an issue https://github.com/dragonflydb/dragonfly/issues/1945 to address this bug

rounaknandanwar commented on src/server/stream_family.cc:
```diff
@@ -812,6 +849,138 @@ OpResult<uint32_t> OpLen(const OpArgs& op_args, string_view key) {
+  return SCG_INVALID_ENTRIES_READ;
+}
+
+// void getConsumerGroupLag(stream* s, streamCG* cg, GroupInfo* ginfo) {
```
Commented out right now because the `entries_read` is not being updated. 
Can start using this in xinfo once `entries_read` updation is fixed.

kostasrim commented:
Hi @rounaknandanwar , it's a packed week but I promise I will take a look in the following days. Will get back to you really soon.

romange commented:
@rounaknandanwar  Hello! Thank you for contributing this PR! Would you like to continue with the code review or you are OK if we take it from here and continue working on it ourselves?

rounaknandanwar commented:
> @rounaknandanwar Hello! Thank you for contributing this PR! Would you like to continue with the code review or you are OK if we take it from here and continue working on it ourselves?

Hi @romange ,
I would like to continue with this PR.
Based on the previous [reply](https://github.com/dragonflydb/dragonfly/pull/1533#discussion_r1325370008) , I need to make some changes. Let me complete those today and then you can start the review.

Also, do take a look at this : https://github.com/dragonflydb/dragonfly/pull/1533#discussion_r1314938024

romange commented:
I did not mean to make it urgent. As long as you continue working on this - it's your PR. We appreciate the time you put into it! 🍻

rounaknandanwar commented:
Hi @romange, Please feel free to start the review. I have completed the changes from my side.

romange commented on src/redis/t_stream.c:
```diff
@@ -1524,6 +1491,41 @@ long long streamEstimateDistanceFromFirstEverEntry(stream *s, streamID *id) {
     return SCG_INVALID_ENTRIES_READ;
 }
 
+#if ROMAN_ENABLE
```
why this has been changed?

romange commented on src/server/stream_family.cc:
```diff
@@ -862,6 +975,207 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
   return result;
 }
 
+vector<Record> streamWithRange(stream* s, streamID start, streamID end, int reverse, size_t count) {
```
please call it `ReadRange` or some other name that starts with a verb. Also please note  our naming convention for functions

rounaknandanwar replied:
Renamed it to `GetStreamRecords`

romange commented on src/server/stream_family.cc:
```diff
@@ -862,6 +975,207 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+    arraylen_cg_pel++;
+  }
+  raxStop(&ri_cg_pel);
+  ginfo->stream_nack_vec = nack_info_vec;
```
std::move(nack_info_vec)

romange commented on src/server/stream_family.cc:
```diff
@@ -862,6 +975,207 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+  ginfo->stream_nack_vec = nack_info_vec;
+}
+
+void getConsumers(stream* s, streamCG* cg, GroupInfo* ginfo, long long count) {
```
same - please order the arguments and the case

romange commented on src/server/stream_family.cc:
```diff
@@ -862,6 +975,207 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+    raxStop(&ri_cpel);
+  }
+  raxStop(&ri_consumers);
+  ginfo->consumer_info_vec = consumer_info_vec;
```
move

rounaknandanwar replied:
Done.

romange commented:
hello @rounaknandanwar , thank you for submitting this PR!  I trust you checked the code - I can not review for correctness fpr such big PR. I gave some comments about style and performance issues. Please note that currently you have conflicts with t_stream.c that must be resolved.

rounaknandanwar commented on src/server/stream_family.cc:
```diff
@@ -854,6 +962,11 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
       ginfo.last_id = cg->last_id;
+      // ginfo.entries_read = cg->entries_read;  // TODO : this returns incorrect value, check.
+      ginfo.entries_read = streamEstimateDistanceFromFirstEverEntry(
+          s, &cg->last_id);  // TODO : check if this is correct.
```
Have removed the TODOs and pushed fixes.

rounaknandanwar commented:
> hello @rounaknandanwar , thank you for submitting this PR! I trust you checked the code - I can not review for correctness fpr such big PR. I gave some comments about style and performance issues. Please note that currently you have conflicts with t_stream.c that must be resolved.

Hi @romange,

Yes, I have verified the correctness by comparing it directly to redis output. I have also gone through their docs to note any edge cases that need to be handled here as well.
Also, I have added tests - those should give you a fair idea of the flow and correctness.

Regarding the comments and conflicts, I have fixed them all.

romange commented on src/server/stream_family.cc:
```diff
@@ -876,6 +911,205 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+  CompactObj& cobj = (*res_it)->second;
+  stream* s = (stream*)cobj.RObjPtr();
+
+  streamCG* cg = streamLookupCG(s, sdsnewlen(group_name.data(), group_name.length()));
```
there is a memory leak here.`sdsnewlen` allocates an sds string that is never freed

rounaknandanwar replied:
Have used the tmp_str1 from shard. Hope that is the correct way to do it across.

romange commented on src/server/stream_family.cc:
```diff
@@ -876,6 +911,205 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+      sinfo.cgroups = group_info_vec;
+    }
+  } else {
+    sinfo.groups = s->cgroups ? raxSize(s->cgroups) : 0;
```
this line already appears at 1032

rounaknandanwar replied:
removed.

romange commented on src/server/stream_family.cc:
```diff
@@ -876,6 +911,205 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
   return result;
 }
 
+vector<Record> GetStreamRecords(stream* s, streamID start, streamID end, int reverse,
```
lets change `int reverse` to `bool reverse` and when calling this function, please pass false/true instead fo `0/1`

rounaknandanwar replied:
done.

romange reviewed (approved):
Thank you for your contribution @rounaknandanwar !
